resnet/main.py

```diff
@@ -0,0 +1,238 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+"""Example script to train a ResNet model on ImageNet."""
+
+import os
+import sys
+
+import torch
+from composer import Trainer
+from composer.algorithms import (EMA, SAM, BlurPool, ChannelsLast, ColOut,
+                                 LabelSmoothing, MixUp, ProgressiveResizing,
+                                 RandAugment, StochasticDepth)
+from composer.callbacks import MemoryMonitor, LRMonitor, SpeedMonitor
+from composer.loggers import ProgressBarLogger, WandBLogger
+from composer.optim import CosineAnnealingWithWarmupScheduler, DecoupledSGDW
+from composer.utils import dist
+from omegaconf import OmegaConf
+
+from data import build_imagenet_dataspec, build_streaming_imagenet_dataspec
+from model import build_composer_resnet
+
+
+def build_logger(name, kwargs):
+    if name == 'progress_bar':
+        return ProgressBarLogger(
+            progress_bar=kwargs.get('progress_bar', True),
+            log_to_console=kwargs.get('log_to_console', True),
+        )
+    elif name == 'wandb':
+        return WandBLogger(**kwargs)
+    else:
+        raise ValueError(f'Not sure how to build logger: {name}')
+
+def log_config(cfg):
+    print(OmegaConf.to_yaml(cfg))
+    if 'wandb' in cfg.loggers:
+        try:
+            import wandb
+        except ImportError as e:
+            raise e
+        if wandb.run:
+            wandb.config.update(OmegaConf.to_container(cfg, resolve=True))
+
+def main(config):
+    if config.grad_accum == 'auto' and not torch.cuda.is_available():
+        raise ValueError('grad_accum="auto" requires training with a GPU, please specify grad_accum as an integer')
+
+    # Divide batch sizes by number of devices if running multi-gpu training
+    train_batch_size = config.train_dataset.batch_size
+    eval_batch_size = config.eval_dataset.batch_size
+    if dist.get_world_size():
+        train_batch_size //= dist.get_world_size()
+        eval_batch_size //= dist.get_world_size()
+
+    # Train dataset
+    print('Building train dataloader')
+    if config.train_dataset.is_streaming:
+        train_dataspec = build_streaming_imagenet_dataspec(
```

Comment:
ya, this block makes me more convinced we should deduplicate the two dataspec builder functions--having separate functions lengthens both the mains training script and data.py

Reply:
Done!